Task 8

iharmonyharper opened the pull request.

iharmonyharper committed:
change filtering

iharmonyharper committed:
Task 8. Custom filters

zverok commented on lib/movie/movie.rb:
```diff
@@ -106,13 +109,16 @@ def pretty_print
     private
 
+    def match_filters?(f, value)
+      value === send(f) || (Array === send(f) ? send(f).include?(value.to_s) : false)
```
Это можно лучше:
1. вынести `send(f)` в переменную, а то повторяется много раз.
2. `Array === ...` — лучше `....is_a?(Array)`
3. `false` в явную как один из результатов условий — обычно признак того что можно упростить:
```ruby
v.is_a?(Array) && v.include?(value.to_s)
```

zverok commented on lib/theaters/netflix.rb:
```diff
@@ -0,0 +1,65 @@
+      @custom_filters ||= {}
+      if from
+        @custom_filters[filter_name] = @custom_filters[from]
+        @custom_filters["#{filter_name}_arg".to_sym] = arg if arg
```
Я бы сказал что в одном хеше хранить и фильтр, и аргумент — нехорошо совсем. И в принципе лучше бы здесь сразу произвести «фильтрующий блок» из исходного фильтра и аргумента, и сохранить его. М?

zverok commented on lib/theaters/netflix.rb:
```diff
@@ -0,0 +1,65 @@
+
+    def show(**filter, &block)
+      customer, default = filter.partition do |k, _v|
+        custom_filters.keys.include?(k) if custom_filters
```
Этот `if` можно убрать если `custom_filters` всегда иметь инициализированным: или в initialize, или через
```ruby
def custom_filters
  @custom_filters ||= {}
end
```

zverok commented on lib/theaters/netflix.rb:
```diff
@@ -0,0 +1,65 @@
+          block = proc { |x| custom_proc.first.call(x, f_value) } unless block_given?
+          block = proc { |x| block.call(x) && custom_proc.first.call(x, f_value) } if block_given?
+        end
+      end
```
Это лучше бы вынести в отдельный метод, и сильно упростить. Там в идеале должно (бы) остаться что-то типа «получили фильтр → применили фильтр» без такого количества `if`-ов и дополнительных переменных.

zverok reviewed (changes requested):
И, к слову, рубокоп надо довести до такого состояния (с помощью исправлений кода и исключений в `.rubocop.yml`), чтобы он был полностью зелёненький :) Это хороший способ «контролировать себя» (папку `spec` можно из проверки исключить — а можно и не исключать, просто добавить пару исключений специально для неё, на размер блока напр.)

iharmonyharper committed:
*Task 8.
- add def custom_filters
- change def match_filters?
- save new filter with arg from the old one.

iharmonyharper committed:
*Task 8.
- move selection of custom filter/block from #show method

Commit:
*Task 8.
- move selection of custom filter/block from #show method
- rubocop rules

zverok commented on .rubocop.yml:
```diff
@@ -0,0 +1,28 @@
+   Enabled: false
+
+AbcSize:
+  Max: 19
```
Это очень много. AbcSize один из самых полезных копов, он действительно показывает алгоритмически проблемные места. В отличие от штук типа длины строки (которые можно считать персональным стилем), его задирать имеет смысл только если точно понятно, что это необходимо. Но лучше — рефакторить и упрощать.

zverok commented on .rubocop.yml:
```diff
@@ -0,0 +1,28 @@
+   Enabled: false
+
+CaseEquality:
+   Enabled: false
```
Лучше задизейблить локально (комментом `#rubocop:disable CaseEquality`) в том единственном месте, где оно действительно оправданно.

zverok commented on .rubocop.yml:
```diff
@@ -0,0 +1,28 @@
+  Max: 105
+
+AccessorMethodName:
+   Enabled: false
```
А вот это зря. Подсказку он давал совершенно верную, `get_movies` действительно неудачное имя. По контексту я бы скорее сказал `create_movies`.

zverok commented on lib/theaters/netflix.rb:
```diff
@@ -0,0 +1,88 @@
+      end
+
+      [filter, block]
+    end
```
Сложновато получилось, тяжело прочесть! И довольно хрупко притом. Я бы сказал, что путь к упрощению — не пытаться сначала как-то преобразовать фильтры, а потом в другом методе использовать, а прямо вот ими и фильтровать:
```ruby
custom, standard = filters.partition{...}
result = custom.inject(@movies) { |res, (name, arg)|
  block = custom[name]
}
result = standard.inject(result) { ... }
```
и т.д.
м? Мне кажется, тут 3-4 выражения всего получится, без кучи всопомогательных методов.

iharmonyharper committed:
*Task 8.
- rubocop rules

iharmonyharper committed:
*Task 8.
- filter_collection(

zverok commented on lib/theaters/netflix.rb:
```diff
@@ -0,0 +1,75 @@
+        custom_filters[filter_name] = proc { |x| custom_filters[from].call(x, arg) } if arg
+      elsif block_given?
+        custom_filters[filter_name] = block
+      end
```
Мне кажется, вот так будет лучше (более DRY):
```ruby
custom_filters[filter_name] =
  if from && arg
    ->(mov) { custom_filters[from].call(mov, arg) }
  elsif from
    custom_filters[from] # хотя этот вариант особого смысла не имеет, может ругаться на него? :)
  elsif block
    block
  else
    # что-то пошло не так, тоже имеет смысл ругаться
  end
```

zverok commented on lib/theaters/netflix.rb:
```diff
@@ -0,0 +1,75 @@
+
+    def filter_collection(**filter, &block)
+      custom, default = filter.partition { |k, _v| custom_filters.keys.include?(k) }.map(&:to_h)
+      filter = default
```
Зачем это переименование?..

zverok commented on lib/theaters/netflix.rb:
```diff
@@ -0,0 +1,75 @@
+        else
+          block = proc { |x| custom_filters[name].call(x, arg) }
+        end
+        res.select(&block).select { |m| m.matches? filter }
```
Здесь получается а) если фильтр с аргументами, то затирается переданный в метод `block` и б) этот `block` хранится во всех итерациях, и постоянно применяется (вреда от того не будет, но выглядит странно) и в) один и тот же `filter` тоже применяется всё время. Не проще ли так:
```ruby
custom.inject(result) do |res, (name, arg)|
  filter = custom_filters[name]
  if filter.arity.zero?
    res.select(&filter)
  else
    res.select { |mov| filter.call(mov, arg) }
  end
end
```
м?.

iharmonyharper committed:
*Task 8.
- custom filters can be simple O_o

zverok commented on lib/theaters/netflix.rb:
```diff
@@ -0,0 +1,65 @@
+        filter = custom_filters[name]
+        res.select { |m| filter.call(m, arg) }
+      end
+    end
```
Ну дык! Лучше же стало, не?.. Только имя бы этому методу какое-то более коммуникативное (откуда get, какую collection?). Скажем `apply_filters`?..

zverok reviewed (changes requested):
А спеки теперь для всех новых фильтров?.. (А по-хорошему, с них бы — спек — и стоило начинать ;))

iharmonyharper committed:
*Task 8.
- rename apply_filters
- add tests

iharmonyharper committed:
*Task 8.
- rubocop

zverok commented on spec/netflix_spec.rb:
```diff
@@ -80,6 +80,57 @@ module Cinematic
+      it 'should be a block' do
+        netflix.define_filter(:new_sci_fi) {{genre: 'Sci-Fi', period: :modern}}
+        expect(netflix.custom_filters[:new_sci_fi]).to be_a_kind_of(Proc)
+      end
```
О. А вот с тестами-то видно, что задание понято не вполне верно. «Блок, который вернёт хеш» — это довольно бессмысленная фича. А надо бы вот так:
```ruby
netflix.define_filter(:something) { |movie| movie.genre == ['Sci-Fi', 'Drama'].sample }
# или скажем
netflix.define_filter(:something_else) { |movie| movie.length < 120 && movie.genre == 'Drama' || movie.genre == 'Sci-Fi' && movie.rating > 9.2 }
```
и т.д.

iharmonyharper committed:
*Task 8.
- tests

zverok reviewed (approved):
:+1:
